File: src/intTest/java/io/javalin/community/ssl/TLSConfigTest.java
package io.javalin.community.ssl;

import io.javalin.Javalin;
import io.javalin.community.ssl.certs.Server;
import okhttp3.*;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import javax.net.ssl.SSLHandshakeException;
import java.net.UnknownServiceException;
import java.util.Arrays;
import java.util.Collections;

import static org.junit.jupiter.api.Assertions.*;

@Tag("integration")
public class TLSConfigTest extends IntegrationTestClass {

    private static String[] substractArray(String[] array, String[] toRemove) {
        return Arrays.stream(array).filter(s -> !Arrays.asList(toRemove).contains(s)).toArray(String[]::new);
    }

    private static OkHttpClient clientWithTLSConfig(TLSConfig config) {

        ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
            .tlsVersions(config.getProtocols())
            .cipherSuites(config.getCipherSuites())
            .build();

        return untrustedClientBuilder()
            .connectionSpecs(Collections.singletonList(spec))
            .build();
    }

    @Test
    @DisplayName("Test that a Modern TLS config does not allow old protocols")
    void testModernConfigWithOldProtocols() {

        String[] protocols = substractArray(TLSConfig.OLD.getProtocols(), TLSConfig.MODERN.getProtocols()); // remove modern protocols from old protocols, so that ONLY unsupported protocols are left

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(TLSConfig.MODERN.getCipherSuites(), protocols));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.MODERN;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old protocols
            assertThrows(SSLHandshakeException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }

    @Test
    @DisplayName("Test that a Modern TLS config does not allow old cipher suites")
    void testModernConfigWithOldCipherSuites() {

        String[] cipherSuites = substractArray(TLSConfig.OLD.getCipherSuites(), TLSConfig.MODERN.getCipherSuites()); // remove modern cipher suites from old cipher suites, so that we can test ONLY the old cipher suites

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(cipherSuites, TLSConfig.MODERN.getProtocols()));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.MODERN;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old cipher suites
            assertThrows(SSLHandshakeException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }

    @Test
    @DisplayName("Test that an Intermediate TLS config does not allow old protocols")
    void testIntermediateConfigWithOldProtocols() {

        String[] protocols = substractArray(TLSConfig.OLD.getProtocols(), TLSConfig.INTERMEDIATE.getProtocols()); // remove intermediate protocols from old protocols, so that ONLY unsupported protocols are left

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(TLSConfig.INTERMEDIATE.getCipherSuites(), protocols));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.INTERMEDIATE;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old protocols
            assertThrows(UnknownServiceException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }


    @Test
    @DisplayName("Test that an Intermediate TLS config does not allow old cipher suites")
    void testIntermediateConfigWithOldCipherSuites() {

        String[] cipherSuites = substractArray(TLSConfig.OLD.getCipherSuites(), TLSConfig.INTERMEDIATE.getCipherSuites()); // remove intermediate cipher suites from old cipher suites, so that we can test ONLY the old cipher suites

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(cipherSuites, TLSConfig.INTERMEDIATE.getProtocols()));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.INTERMEDIATE;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old cipher suites
            assertThrows(SSLHandshakeException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }

    @Test
    @DisplayName("Test an Intermediate TLS config works with TLSv1.3")
    void testIntermediateConfigWithTLSv13() {

        ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.RESTRICTED_TLS)
            .tlsVersions(TlsVersion.TLS_1_3)
            .build();

        OkHttpClient client = untrustedClientBuilder()
            .connectionSpecs(Collections.singletonList(spec))
            .build();


        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.INTERMEDIATE;
        }).start()) {
            //Should work with TLSv1.3
            try (Response response = client.newCall(new Request.Builder().url(https).build()).execute()) {
                assertEquals(200, response.code());
                assertEquals(TlsVersion.TLS_1_3,response.handshake().tlsVersion());
            } catch (Exception e) {
                e.printStackTrace();
                fail(e);
            }
        }
    }
}

File: src/intTest/java/io/javalin/community/ssl/TLSConfigTest.kt
package io.javalin.community.ssl

import io.javalin.community.ssl.certs.Server
import okhttp3.ConnectionSpec
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.TlsVersion
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import java.net.UnknownServiceException
import javax.net.ssl.SSLHandshakeException

@Tag("integration")
class TLSConfigTest : IntegrationTestClass() {
    @Test
    @DisplayName("Test that a Modern TLS config does not allow old protocols")
    fun testModernConfigWithOldProtocols() {

        val protocols = TLSConfig.OLD.protocols.subtract(TLSConfig.MODERN.protocols.asIterable())

        // remove modern protocols from old protocols, so that ONLY unsupported protocols are left
        val client = clientWithTLSConfig(TLSConfig(TLSConfig.MODERN.cipherSuites, protocols.toTypedArray()))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.MODERN
        }.start().use { _ ->
            //Should fail with SSLHandshakeException because of the old protocols
            Assertions.assertThrows(SSLHandshakeException::class.java) {
                client.newCall(
                    Request.Builder().url(
                        https
                    ).build()
                ).execute()
            }
        }
    }

    @Test
    @DisplayName("Test that a Modern TLS config does not allow old cipher suites")
    fun testModernConfigWithOldCipherSuites() {
        val cipherSuites = TLSConfig.OLD.cipherSuites.subtract(TLSConfig.MODERN.cipherSuites.asIterable())
        // remove modern cipher suites from old cipher suites, so that we can test ONLY the old cipher suites
        val client = clientWithTLSConfig(TLSConfig(cipherSuites.toTypedArray(), TLSConfig.MODERN.protocols))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.MODERN
        }.start().use { _ ->
            //Should fail with SSLHandshakeException because of the old cipher suites
            Assertions.assertThrows(SSLHandshakeException::class.java) {
                client.newCall(
                    Request.Builder().url(
                        https
                    ).build()
                ).execute()
            }
        }
    }

    @Test
    @DisplayName("Test that an Intermediate TLS config does not allow old protocols")
    fun testIntermediateConfigWithOldProtocols() {
        val protocols = TLSConfig.OLD.protocols.subtract(TLSConfig.INTERMEDIATE.protocols.asIterable())
        // remove intermediate protocols from old protocols, so that ONLY unsupported protocols are left
        val client = clientWithTLSConfig(TLSConfig(TLSConfig.INTERMEDIATE.cipherSuites, protocols.toTypedArray()))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.INTERMEDIATE
        }.start().use { _ ->
            //Should fail with SSLHandshakeException because of the old protocols
            Assertions.assertThrows(UnknownServiceException::class.java) {
                client.newCall(
                    Request.Builder().url(https).build()
                ).execute()
            }
        }
    }

    @Test
    @DisplayName("Test that an Intermediate TLS config does not allow old cipher suites")
    fun testIntermediateConfigWithOldCipherSuites() {
        val cipherSuites = TLSConfig.OLD.cipherSuites.subtract(TLSConfig.INTERMEDIATE.cipherSuites.asIterable())
        // remove intermediate cipher suites from old cipher suites, so that we can test ONLY the old cipher suites
        val client = clientWithTLSConfig(TLSConfig(cipherSuites.toTypedArray(), TLSConfig.INTERMEDIATE.protocols))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.INTERMEDIATE
        }.start().use { _ ->
            //Should fail with SSLHandshakeException because of the old cipher suites
            Assertions.assertThrows(SSLHandshakeException::class.java) {
                client.newCall(
                    Request.Builder().url(
                        https
                    ).build()
                ).execute()
            }
        }
    }

    @Test
    @DisplayName("Test an Intermediate TLS config works with TLSv1.3")
    fun testIntermediateConfigWithTLSv13() {
        val spec: ConnectionSpec = ConnectionSpec.RESTRICTED_TLS
        val client: OkHttpClient = untrustedClientBuilder().connectionSpecs(listOf(spec))
            .build()
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.INTERMEDIATE
        }.start().use { _ ->
            //Should work with TLSv1.3
            try {
                client.newCall(Request.Builder().url(https).build()).execute().use { response ->
                    Assertions.assertEquals(200, response.code)
                    Assertions.assertEquals(TlsVersion.TLS_1_3, response.handshake!!.tlsVersion)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                Assertions.fail<Any>(e)
            }
        }
    }

    companion object {

        private fun clientWithTLSConfig(config: TLSConfig): OkHttpClient {


            val spec = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
                .cipherSuites(*config.cipherSuites)
                .tlsVersions(*config.protocols)
                .build()

            return listOf(spec).let {
                OkHttpClient.Builder()
                    .connectionSpecs(it)
                    .build()
            }
        }
    }
}
